torch/csrc/jit/passes/onnx/shape_type_inference.cpp

```diff
@@ -1374,13 +1392,88 @@ void ONNXShapeTypeInference(
 
 } // namespace
 
+std::pair<bool, bool> IsShapeReliable(Node* n) {
+  auto reliable = true;
+  auto complete = true;
+  for (auto input : n->inputs()) {
+    reliable &=
+        ConstantValueMap::GetTypeReliable(input->debugName()).value_or(false);
+    if (auto pt = input->type()->cast<TensorType>()) {
+      if (!pt->sizes().isComplete()) {
+        complete = false;
+      }
+    }
+  }
+  return std::make_pair(reliable, complete);
+}
+
+// There is no need to put onnx type here, but we need this
+// for some legacy tests when onnx_shape_inference=False.
+static std::unordered_set<std::string> nodeTypeReliableForTracer = {
+    "prim::ListConstruct",
+    "onnx::Cast",
+    "onnx::Constant",
+    "com.microsoft::Gelu"};
+
+void UpdateReliable(
+    torch::jit::Value* output,
+    const std::pair<bool, bool>& inferred_type_reliable) {
```

Comment:
it appears `inferred_type_reliable` is (always?) achieved from `IsShapeReliable(output->node())`. Can we remove this argument and get the info inside this call then?

If there is a performance concern, would it be helpful to have overload of `UpdateReliable(Node* n)`?

Reply:
Right, there is a performance concern, because we only want to call `IsShapeReliable` once for all outputs. Just create the overload for now, because there are only two places will use `UpdateReliable(Node* n)`. May need refactor after early termination logic change.
For other places, it is about (1) update input reliable for `prim::ListConstruct`, (1) already in loop.